### PR TITLE
JSSpecCompiler: Road to testing (part 2/3)

### PR DESCRIPTION
**Pull request by DanShaders**
~~This is definitely not the next JSSpecCompiler PR, so I'm not breaking the promise :)~~

More seriously, I've reasoned `25 files changed, 608 insertions(+), 56 deletions(-)` is a little bit too much for a single PR, so I decided to split CLI and test runner itself into different PRs.

Note that in the state between these two PRs, there won't be any way to tell the compiler to output something.

### CODE REVIEW

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/CppASTConverter.cpp [hunk 1]
```diff
@@ -175,4 +177,32 @@ Tree CppASTConverter::as_possibly_empty_tree(Cpp::Statement const* statement)
...
+}
+
+CppParsingStep::~CppParsingStep() = default;
+
+void CppParsingStep::run(TranslationUnitRef translation_unit)
+{
+    auto filename = translation_unit->filename;
+
+    auto file = Core::File::open_file_or_standard_stream(filename, Core::File::OpenMode::Read).release_value_but_fixme_should_propagate_errors();
+    m_input = file->read_until_eof().release_value_but_fixme_should_propagate_errors();
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/CppASTConverter.cpp:192]: after the run function, this m_input variable will be a dangling reference, as you don't hold onto the file object.

Should it really be a member?

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/CompilationPipeline.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+public:
+    CompilationStep(StringView name)
+        : m_name(name)
+    {
+    }
+
+    virtual ~CompilationStep() = default;
+    virtual void run(TranslationUnitRef translation_unit) = 0;
+
+    StringView m_name;
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/CompilationPipeline.h:26]: public member vars in `class` types are sad. Should this be a struct? or can we encapsulate the class's data members.

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/SpecParser.cpp [hunk 3]
```diff
@@ -173,4 +176,31 @@ ParseErrorOr<void> SpecFunction::parse_definition(XML::Node const* element)
...
+}
+
+SpecParsingStep::~SpecParsingStep() = default;
+
+void SpecParsingStep::run(TranslationUnitRef translation_unit)
+{
+    auto filename = translation_unit->filename;
+
+    auto file = Core::File::open_file_or_standard_stream(filename, Core::File::OpenMode::Read).release_value_but_fixme_should_propagate_errors();
+    m_input = file->read_until_eof().release_value_but_fixme_should_propagate_errors();
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/SpecParser.cpp:191]: same question here about this dangling reference.

Future refactor might try to use MappedFile instead.

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/CppASTConverter.cpp (hunk 1 above)
Comment (DanShaders) [reply Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/CppASTConverter.cpp:192]: `m_input` is a `ByteBuffer`, which is an owner of all string views we have in ASTs (and this is actually the primary reason why I keep all of the compilation step classes until compilation finishes)

**Comment by DanShaders** [comment]
Changes since review:
- Made `CompilationStep::m_name` private
- Changed implementation of (currently unused) `CompilationPipeline::for_each_step_in`

**Pull request merged**